XLS-34d-paychan-escrow-for-tokens/README.md

```diff
@@ -0,0 +1,252 @@
+<pre class="notranslate">
+Title:       <b>Token-Enabled Escrows and Payment Channels</b>
+Type:        Draft
+PR:          <a href="https://github.com/XRPLF/rippled/pull/4396">4396</a>
+Author:      Richard Holland, Xaman <a href="https://github.com/RichardAH/">(RichardAH)</a>
+             Denis Angell, Transia <a href="https://github.com/dangell7/">(dangell7)</a>
+Affiliation: <a href="https://xrplf.org" rel="nofollow">XRPLF</a>
+</pre>
+
+> This proposal, XLS34d, replaces [XLS27d](https://github.com/XRPLF/XRPL-Standards/discussions/72)
+
+# Token-Enabled Escrows and Payment Channels
+
+The proposed amendment to the XRPL protocol, `PaychanAndEscrowForTokens`, would introduce changes to the ledger objects, transactions, and rpc methods to enable `Escrows` and `PayChannels` to use Trustline balances.
+
+## 1. Problem Statement
+
+The XRPL supports several types of on-ledger negotiable Instruments, namely: `Escows`, `PayChannels` and `Checks`. While each of these instruments is implemented as a first-class on-ledger object, only the `Check` object supports the use of Trustline balances. `PayChannels` and `Escrows` support only the native asset XRP. This limitation is a barrier to wider-spread use of these Instruments for reasons including:
+
+- Regulatory compliance.
+- Unwillingness to hold a counter-party-free asset (i.e. XRP).
+- Volatility and exchange-rate risk.
+
+## 2. Constraints
+
+The amendment has the following constraints:
+
+- Allow `Escrows` to use Trustline balances.
+- Allow `PayChannels` to use Trustline balances.
+- Allow Issuers to retain freeze and authorization control over their Issued tokens even when they are locked into Instruments.
+- Avoid unnecessary multiplication of on-ledger entities.
+- Avoid unnecessarily burdening the ledger computationally.
+
+## 3. Implementation
+
+An Instrument can only be created over a Trustline balance if both parties (source and destination):
+
+- Are Authorised to hold the Trustline,
+- Do not have a frozen Trustline for the issuer/currency pair, and
+- Can ripple the balance through its Issuer.
+
+The destination may not have a Trustline yet. This is not a bar to creating the Instrument provided that a hypothetical Trustline created at the same time the Instrument is created, from the destination account to the Issuer, for the relevant currency, would meet the above requirements.
+
+Settlement of an Instrument (e.g. `EscrowFinish`, `PaymentChannelClaim`) carries the same semantics as crossing an offer: If no Trustline exists on the destination side, then a Trustline with zero limit is created and the tokens (balance) are placed into that line. However creation of a Trustline requires the transaction to be signed by the receiving account (because it creates an on-ledger object and increases the Owner Count and thus the XRP reserve locked by the destination account.)
+
+### 3.1. New Field: `LockedBalance`.
```

Comment:
I know I'm late to the party, but one thing I haven't been clear on, and I don't think any of these specs have answer is: What exactly are the `LockedBalance` and `LockCount` fields trying to solve?

In other words, if one were to do a "naive" implementation, and just move the balance from the source trustline to the Escrow object when creating it, and then move it from the Escrow to the source or destination when cancelling or finishing, respectively, what would be broken?

Reply:
When we do PayChan or Escrow for XRP the XRP is "moved" from the account to the LE. In this implementation the money is still in the account. It is `Locked`. The `LockedBalance` is the amount locked in the trustline. The count is the number of locks on the line. So if you had 2 escrows and 3 paychans the `LockCount` would be 5.

Technically doing it "naive" way that would go against the exact thing seelabs pointed out.

"The biggest concern I have with this proposal is the way balances are now represented in the ledger."

We don't actually do that. The money stays in the account however the amount is not transferable because of functionality we implemented.

Reply:
@dangell7 Thanks for that explanation.